hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieBaseRelation.scala

```diff
@@ -122,29 +122,39 @@ abstract class HoodieBaseRelation(val sqlContext: SQLContext,
     optParams.get(DataSourceReadOptions.TIME_TRAVEL_AS_OF_INSTANT.key)
       .map(HoodieSqlCommonUtils.formatQueryInstant)
 
+  /**
+   * NOTE: Initialization of teh following members is coupled on purpose to minimize amount of I/O
+   *       required to fetch table's Avro and Internal schemas
+   */
   protected lazy val (tableAvroSchema: Schema, internalSchema: InternalSchema) = {
-    val schemaUtil = new TableSchemaResolver(metaClient)
-    val avroSchema = Try(schemaUtil.getTableAvroSchema) match {
-      case Success(schema) => schema
-      case Failure(e) =>
-        logWarning("Failed to fetch schema from the table", e)
-        // If there is no commit in the table, we can't get the schema
-        // t/h [[TableSchemaResolver]], fallback to the provided [[userSchema]] instead.
-        userSchema match {
-          case Some(s) => convertToAvroSchema(s)
-          case _ => throw new IllegalArgumentException("User-provided schema is required in case the table is empty")
-        }
+    val schemaResolver = new TableSchemaResolver(metaClient)
+    val avroSchema: Schema = schemaSpec.map(convertToAvroSchema).getOrElse {
+      Try(schemaResolver.getTableAvroSchema) match {
+        case Success(schema) => schema
+        case Failure(e) =>
+          logError("Failed to fetch schema from the table", e)
+          throw new HoodieSchemaException("Failed to fetch schema from the table")
+      }
     }
-    // try to find internalSchema
-    val internalSchemaFromMeta = try {
-      schemaUtil.getTableInternalSchemaFromCommitMetadata.orElse(InternalSchema.getEmptyInternalSchema)
-    } catch {
-      case _: Exception => InternalSchema.getEmptyInternalSchema
+
+    val schemaEvolutionEnabled: Boolean = optParams.getOrElse(DataSourceReadOptions.SCHEMA_EVOLUTION_ENABLED.key,
```

Comment:
we cannot rely on this config,  user maybe forget set this config which can lead a incorrectly result

Reply:
@xiarixiaoyao : A tangential question trying to understand something. can you help me understand the interplay here or how the config us used. for the first time on read path, if someone enables this config. After that, all read and write paths will rely on presence of InternalSchema is it? or is it that write path will always rely on the schema evolution enabled config and read path will rely on the presence of internal schema?

Reply:
@xiarixiaoyao we have to make this config mandatory now on the read-path as well -- without it there's no way for us to know whether we should be reading Hudi table via V1 or V2 API (currently we fallback to V1 for all use-cases except schema evolution to make sure there's no regression in performance in Spark SQL compared to 0.10)

Reply:
@nsivabalan Sorry for the late reply
if someone enables SCHEMA_EVOLUTION_ENABLED and do some DDL operation
only DDL operation need to enable SCHEMA_EVOLUTION_ENABLED 
all read and write paths will rely on presence of InternalSchema， on the read path, we just check for presence of internal schema，

Reply:
@alexeykudinkin   sure you are right.   
we has more efficient method to judge whether schema evolution has been enabled， no need to parse InternalSchema from commitData,  We can just judge Whether there is a historyschema file exist in the tablepath/.hoodie/.schema directory. This operation costs very little.
Don't worry about this. Once this PR is integrated, I will follow up a pr